### DISCUSSION TITLE
When `winget upgrade --all` should be restricted

### DISCUSSION THREAD

**Discussion by denelon** [microsoft/winget-cli#2407] [Q&A]
There are a few classes of packages that users would like to exclude from `winget upgrade --all`.

Do the following classes and assumptions resonate, or should the intended behavior be modified?

## Version Unknown
The installed package doesn't report its version. In manifests, the version "Unknown" is logically treated as the lowest possible version.
When users run `winget upgrade` these packages are not displayed as available to be upgraded, but they can be displayed by passing "--include-unknown".  Users may also run `winget upgrade --all --include-unknown` and the package manager will attempt to upgrade them. Generally, these package installers are considered suboptimal since the Windows Package Manager cannot reason about whether an available manifest represents an upgrade.

## Packages that automatically upgrade
These packages like Microsoft.VisualStudioCode and Microsoft.EdgeWebView2Runtime upgrade themselves. This may happen in the background as with Edge Web View 2 Runtime, or when the program is launched or closed as with Visual Studio Code. The manifest key "requiresExplicitUpgrade: true" provides the hint to the Windows Package Manager that these packages should be excluded from `winget upgrade --all`. A user may still run `winget ugpgrade <package>` to upgrade these packages individually.

When users run `winget upgrade` these packages will be displayed separately to inform the user that they will be excluded from `winget upgrade --all` if a newer version is available.

## Excluded or Pinned packages
The packages that automatically upgrade can be thought of as excluded in the sense they are excluded from `winget upgrade --all`. 

Two separate features have been requested with subtle differences in intent:
* #476 
* #2339

### Pinned Packages
Pinning a package is perceived as an active intent a user has provided to ensure the Windows Package Manager does not upgrade to a newer version of the package, and to force additional action on the user to "unpin" a package before allowing upgrades. There are caveats with pinning to a major, minor, or patch level version to allow an approved subset of version upgrades. When users run `winget upgrade` these packages will be displayed separately to inform the user that they will be excluded from `winget upgrade --all` if a newer version is available.

A user would not be able to `winget upgrade <package>` to a version beyond the pin. A user would be required to `winget unpin <package>` first, or to pass the "--reset-pin" (`winget upgrade <package> --reset-pin`) argument to upgrade beyond the pinned version.

The Windows Package Manager is not able to prevent other means of upgrading packages. Any package with the key "requiresExplicitUpgrade: true" set in the manifest would not be "pinnable". In the event a package that had been pinned is upgraded beyond the pinned version outside of the Windows Package Manager, the Windows Package Manager would inform the user via `winget list`, `winget upgrade`, or `winget show <package>`. The user would also be informed that the pin had been removed due to this upgrade.

### Excluded Packages
Excluding a package is perceived as an active intent a user has provided to ensure the Windows Package Manager does not upgrade to a newer version of the package, but not to force additional action on the user to "include" a package before allowing upgrades. Running `winget upgrade <package>` would upgrade the package to the latest available version. The corollary `winget include` command could be used to remove the exclusion of these packages (even the ones that automatically upgrade). When users run `winget upgrade` these packages will be displayed separately to inform the user that they will be excluded from `winget upgrade --all` if a newer version is available.

**Comment by sundhaug92** [comment]
Maybe license change?

**Reply by sundhaug92** [reply]
As in maybe the proposed + license change

**Reply by denelon** [reply]
We do have a mechanism to track source agreement versions.

We do not yet have a mechanism to track license changes or agreement changes for installed packages with a version number we don't have a manifest for.

Example:

I installed Contoso's Awesome App version 1.0 from a download.

The first manifest for Contoso's Awesome App version 2.0 containing "license: MIT" is approved.

During an upgrade beyond version 2.0 (assuming all versions are submitted), we detect a newer version is available, and we would know the license at that point in time for that manifest.

We could assume that the first time the App is installed with a manifest that the license may have changed. Any time we perform an upgrade using a manifest, we could compare the previous version (assuming we have a manifest, and the license hasn't been modified).

**Comment by RichiCoder1** [comment]
I admit, is the difference between Exclude and Pin enough to justify it? Would a `--force` override on `upgrade` be good enough to cover the `pin`-but-allow-manual-upgrades case?

**Reply by denelon** [reply]
That's what I'm trying to get feedback on. Do developers want "pin" to intentionally mean "force me to make a conscious decision/force me to take additional action" when running `winget upgrade <package>`? Does it make sense to have something like:

* `winget upgrade --all --include-excluded`
* `winget upgrade --all --include-pinned --reset-pin`
* `winget upgrade --all --include-pinned --remove-pin`

For upgrade on excluded:
* `winget upgrade <package>`

For upgrade on pinned:
* `winget upgrade <package> --reset-pin`

or unpin first then repin again aftwards?

This is really about the intended UX.

Would `--exclude` logically give developers what they want?

What kind of warnings do developers want if something was changed and WinGet could detect it?
Should the default be to constantly inform, and a setting could hide the warnings?

Maybe "--exclude" is sufficient for now, and we can invest more deeply into pinning behaviors in the future.

**Comment by florelis** [comment]
> Excluding a package is perceived as an active intent a user has provided to ensure the Windows Package Manager does not upgrade to a newer version of the package, but not to force additional action on the user to "include" a package before allowing upgrades. [...] When users run winget upgrade these packages will be displayed separately to inform the user that they will be excluded from winget upgrade --all if a newer version is available.

I liked the interpretation of the difference between 'exclude' and 'pin' from @RichiCoder1:
> [pin is "I want to track this package, but not upgrade it" while exclude would be "I don't want to have winget recognize this package at all". E.g. doesn't show up in winget export and list.](https://github.com/microsoft/winget-cli/issues/2339#issuecomment-1185946061) 

It makes enough of a difference to warrant having both concepts. With that interpretation, I would not show the excluded packages on `winget upgrade` and would require re-including it before it can be upgraded by us.

---

I have another suggestion for pinning somewhat trying to unify what I've heard from @JohnMcPMS  and @denelon :

We could have two types of pinning:
"strict pinning" - meaning "I want to have this exact version of the package", where "exact version" can also be a range/blob like 1.*
"loose pinning" - meaning "I don't care too much about the exact version, I just want to stop it from updating unless I want to", for example, it updates more often than I'd like, or the updates always mess up my settings

For a strict pin:
* `winget upgrade <package>` fails but `winget upgrade <package> -v 1.2.3` succeeds and sets the pin to that, because if we want a specific version, we probably will also want a specific version after updating
* `winget upgrade --all` updates if and only if there is a version matching the pattern, but there is no argument for updating all pinned
* `winget upgrade` shows the update as normal if it matches the version pattern, or in a separate table if it is not (not sure what to do if both are available...)

For a loose pin:
* `winget upgrade <package>` always works
* `winget upgrade --all` doesn't touch them, but we can add a `--include-...` flag to update them
* `winget upgrade` shows available updates in a separate table
* Equivalent to `RequireExplicitUpgrade: true`

For how to pin, it could be done with `winget install|upgrade --pin <package>` or `winget pin <package>`, and choose loose/strict based on whether a version was included. `winget pin <package> -v ...` would fail if the installed version doesn't match.

This is similar to what @denelon  said changing "pinned" by "strictly pinned" and "excluded" by "loosely pinned" and removing the need to unpin or pass a flag for "strictly pinned" (replaced by specifying version)

**Reply by denelon** [reply]
Nice, I think we're getting somewhere better with this approach.

**Reply by denelon** [reply]
We'll need some fairly extensive tests to cover all the strange instances where a newer version beyond the pin shows up, and then at a later point in time a version that includes the pinned range shows up. The package would logically switch between the sets of packages that would get updated and the "informational" set that wouldn't with `winget upgrade --all`

I'm also wondering about the sequencing to support the minimum OS version problem we encountered with Microsoft.WindowsTerminal.
* #2134
and:
* #2209

**Reply by RichiCoder1** [reply]
As an MVP, and with the proposal above I assume there wouldn't be a concept of ranges. Pinning wouldn't even technically pin to a specific version, so much as exclude it from upgrade logic by default no?

Creating dependency ranges via WinGet seems to indeed open a lot of useful, but very complicated situations like #163 that maybe are worth considering _after_ a first pass? A lot of people just want fixed-version flavors of pinning.

Sorry if I'm misunderstand (and thanks @lechacon for summarizing so well!)

**Reply by denelon** [reply]
@RichiCoder1 we've gotten asks for a handful of scenarios. You are correct with the complicated situations. I'm trying to make a determination about what we could/should implement as a first pass as opposed to trying to implement everything. 

Something like "exclude" to just keep something out of the default `winget upgrade --all` scenario is fairly straightforward. I do like the concept of a "loose pin" rather than another command, but I'm not sure if it would be clear to users.

Maybe "exclude" could be an alias to "pin" and their corollaries "include" and "unpin".

The question in my mind is whether that is something that makes sense, and if it's necessary to have a `winget pin <package>` to lock out both `winget upgrade --all` and `winget upgrade <package>` in which case the user would need to `winget unpin <package>` first or additional arguments as @lechacon suggested.

For me, it comes down to intent.

Is the user saying:
* winget ignore this package
  - because I'll manage it with another tool
* winget keep me patched on this minor version
   - because I don't want the next minor version
* winget keep me on this major version
  - because I don't want the next version with breaking changes
* winget keep me in this range of versions
  - because it doesn't support side by side installation
  - because it doesn't honor semantic versioning
* winget make me perform an extra step to upgrade this package
  - because I don't want break something I might not be paying attention to
* winget tell me if a package was upgraded/modified without your knowledge
  - because I want to be informed

Dependencies will open up a whole host of these kinds of concerns

**Reply by denelon** [reply]
I could also see another use case:

* winget exclude this version because of a security risk or because I know it's broken.

**Reply by RichiCoder1** [reply]
In at least both `scoop` and `chocolatey`'s sense, the primary cases I've seen there have been:
* WinGet ignore this package because I'll manage it with another tool / it'll autoupdate
* WinGet make me perform an extra step

It is worth calling out that they both require explicit `pin add/hold` and `pin remove/unhold` in choco and scoop. 

I'll be honest, the other reason I'm nervous about even considering version ranges is also because I don't how much of the Windows ecosystem even properly complies with SemVer/Major-Minor-Patch in a sense that would make ranges non-surprising behavior.

I think this is also maybe me not seeing cases too, now that I think about it. I'm mostly considering WinGet's behavior as a developer. As a sysadmin/SRE perspective where I'm DSC-ing machines, I could see the case for passing ranges to WinGet. In those cases though, I'd expect it to be an explicit intent kind of thing where I'm either passing a version range to `install/upgrade` or using something like an import file to declaratively define ranges. So, I almost consider ranges an entirely seperate concept from pins, though they certainly have overlap as I imagine there's likely developers who'd want to automatically have patched versions of a runtime for example.

Exclude in the sense of "I want to explicitly exclude this package from an upgrade/install during that operation" I'd expect to only exist as a parameter ala `--exclude packages[]`.

I think I'd agree that I'd find "loose pins" confusing and a possible footgun. I'd much rather have a `winget upgrade <id> --upgrade-pinned` or require `winget unpin <id> && winget upgrade <id> && winget pin <id>` to be explicit.

The "a package has drifted from its installed version" is also something I'd not considered. That's kind of a fascinating case. That could be done via pin-ing (assume winget tracks pinning metadata), but that's something I'd expect to be used in combination with `list` and `--json` ala `winget list --show-pin-drift --json` (pardon my terrible naming)

**Reply by denelon** [reply]
I've never heard of "footgun" before, but it's so perfect!

Naming is hard! I just quoted it https://github.com/microsoft/winget-cli/issues/2421#issuecomment-1206629715

Yes, there are many different dimensions/perspectives to consider which is why I wanted to get as much feedback as possible in this problem space.

Ranges are particularly tricky. I can certainly use VC++ Runtime as my current best example.

**Reply by RichiCoder1** [reply]
> I've never heard of "footgun" before, but it's so perfect!

I do a lot of developer tooling and developer advocacy, and it's one of my favorite turns of phrase along with "pit of success" :)

> Ranges are particularly tricky. I can certainly use VC++ Runtime as my current best example.

Yup, in my experience _most_ Microsoft installers are well behaved in this regard. Outside Microsoft though 😅...

**Reply by ddaymn** [reply]
Here are a couple of examples of this issue that I run into when using upgrade --all

JetBrains dotTrace 2022.3.2                               JetBrains.ReSharper  > 2022.2.2       2022.3.1     winget
Windows Software Development Kit - Windows 10.0.19041.685 Microsoft.WindowsSDK < 10.0.22000.832 10.0.22621.2 winget
JetBrains ETW Host Service (x64)                          JetBrains.ReSharper  > 2022.2.2       2022.3.1     winget
Windows Software Development Kit - Windows 10.0.22000.832 Microsoft.WindowsSDK 10.0.22000.832   10.0.22621.2 winget

The Jetbrains falls into the auto updating application and the latest is installed so a way to exclude that. would be nice.
The Windows SDK falls more into what I think may be a different case than discussing.

I have that version installed
winget list WindowsSDK
Name                                                      Id                   Version      Source
---------------------------------------------------------------------------------------------------
Windows Software Development Kit - Windows 10.0.22621.755 Microsoft.WindowsSDK 10.0.22621.2 winget

So when I run upgrade --all it still downloads the new SDK tries to install and fails because it already is installed.  This is a case where it would need to be an exclued this package by version?  May be just a way to exclude passed in by a file like rsync does.  That way you have to explicitly define something to get that behavior and a user knows where this list is.
